### DISCUSSION TITLE
Creating custom dataset from list of networkx graphs

### DISCUSSION THREAD

**Discussion by AdarshMJ** [pyg-team/pytorch_geometric#4337] [Q&A] [answered]
Hi! So I am working on converting a list of networkx graphs to PyG graphs to train a graph classification framework. I observed that when I convert PyG ---> NetworkX ---> PyG again the labels are not there. If I use 

`from torch_geometric.data import Data
from torch_geometric.loader import DataLoader

data_list = [Data(...), ..., Data(...)]
loader = DataLoader(data_list, batch_size=32)`

I get an error while training AttributeError: 'list' object has no attribute 'num_node_features'

How do I convert the dataset appropriately? Any insights are welcome. Thanks again!

![Screenshot 2022-03-24 at 18 29 45](https://user-images.githubusercontent.com/24205200/159975650-d6fc9f4f-ee75-48da-b832-c8380f37cd4f.png)

**Comment by rusty1s** [comment, marked as answer]
You might need to specify the `node_attrs` and `edge_attrs` arguments in the `to_networkx` call.

**Reply by AdarshMJ** [reply]
I am sorry I am not sure I fully understand, how do i extract the node_attrs and edge_attrs? 
If I do this -
`from torch_geometric.utils.convert import to_networkx, from_networkx
 G = to_networkx(data, node_attrs=['x'],edge_attrs=['edge_attr'])
 print(nx.info(G))`
Not sure what I should supply for edge_attr since this comes originally from the PyG graph that I have.

**Reply by rusty1s** [reply]
Since in your example, you do not have any edge features, there is no need to specify `edge_attr`. The following works for me:
```python
from torch_geometric.datasets import TUDataset
from torch_geometric.utils import from_networkx, to_networkx

dataset = TUDataset('/tmp/TU', name='ENZYMES')
data = dataset[0]

print(data)
G = to_networkx(data, node_attrs=['x'])
data = from_networkx(G)
print(data)
```
```
Data(edge_index=[2, 168], x=[37, 3], y=[1])
Data(x=[37, 3], edge_index=[2, 168])
```

Furthermore, I just added support for conversion of graph-level attributes as well (https://github.com/pyg-team/pytorch_geometric/pull/4343), such that `to_networkx(data, node_attrs=['x'], graph_attrs=['y'])` works as well.

**Reply by AdarshMJ** [reply]
Thank you so much!

**Reply by AdarshMJ** [reply]
Hi! So I am trying to run graph classification on the QM9 dataset, the linear layer uses -

`self.lin = Linear(hidden_channels, dataset.num_classes)` 
The batch size is set to 64.  The hidden_channels = 64 and num_classes=19.

While calculating the loss, I get an error -
RuntimeError: The size of tensor a (64) must match the size of tensor b (19) at non-singleton dimension 1

Any idea how to resolve this?

**Reply by rusty1s** [reply]
Are you sure you apply `self.lin` during the `forward` pass of the model?

**Reply by AdarshMJ** [reply]
Hi Yes. Im following this PyG Graph classification tutorial - https://colab.research.google.com/drive/1I8a0DfQ3fI7Njc62__mVXUlcAleUclnb?usp=sharing

I have only changed the dataset to QM9 from the default Proteins.

**Reply by rusty1s** [reply]
I see. Please note that `QM9` is a regression dataset, and so you will need to make use of `torch.MSELoss` rather than `CrossEntropyLoss` for training.

**Reply by AdarshMJ** [reply]
Okay that makes sense. I will make the change. Thank you!
